### PR TITLE
Add nested and object fields to field capabilities response

### LINKED ISSUE #33237: Field caps: return intermediate fields not just leaves
Currently fieldCaps API returns only the leave fields without any info on 'compound' field such as `object` or `nested` or `join`.
To wit, given an inner field, say `a.b` it is impossible to determine what type `a` is.

### PR DESCRIPTION
**Pull request by jimczi**
This commit adds nested and object fields to the field capabilities response.

Closes #33237

### CODE REVIEW

**Comment by jimczi** [comment]
\cc @costin

**Comment by costin** [comment]
FTR, I've used this PR in SQL and it behaved as expected.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java [hunk 1]
```diff
@@ -86,6 +87,20 @@ protected FieldCapabilitiesIndexResponse shardOperation(final FieldCapabilitiesI
...
                     responseMap.put(field, fieldCap);
+                } else {
+                    continue;
+                }
+                // add nested and object fields
+                int dotIndex = ft.name().indexOf('.');
+                while (dotIndex > -1) {
+                    String parentField = ft.name().substring(0, dotIndex);
+                    ObjectMapper mapper = mapperService.getObjectMapper(parentField);
+                    if (mapper != null) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java]: I was wondering whether this condition may ever be false, maybe mention sub fields in a comment? Should we also break as soon as `responseMap.containsKey(parentField)` is true?

**Review by jpountz — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java]: > Should we also break as soon as responseMap.containsKey(parentField) is true?

If we start from the leaf, yes, I pushed 
https://github.com/elastic/elasticsearch/pull/33803/commits/235df36377ff4c2e171c451bc748097f0ad9b866 to reverse the iteration order.

**Pull request merged**